x-pack/plugins/siem/public/containers/case/use_get_case_user_actions.tsx

```diff
@@ -10,52 +10,128 @@ import { useCallback, useEffect, useState } from 'react';
 import { errorToToaster, useStateToaster } from '../../components/toasters';
 import { getCaseUserActions } from './api';
 import * as i18n from './translations';
-import { CaseUserActions, ElasticUser } from './types';
+import { CaseExternalService, CaseUserActions, ElasticUser } from './types';
+import { parseString } from './utils';
+import { CaseFullExternalService } from '../../../../case/common/api/cases';
+
+interface CaseService extends CaseExternalService {
+  firstPushIndex: number;
+  lastPushIndex: number;
+  hasDataToPush: boolean;
+}
+
+export interface CaseServices {
+  [key: string]: CaseService;
+}
 
 interface CaseUserActionsState {
+  caseServices: CaseServices;
   caseUserActions: CaseUserActions[];
-  firstIndexPushToService: number;
   hasDataToPush: boolean;
-  participants: ElasticUser[];
-  isLoading: boolean;
   isError: boolean;
-  lastIndexPushToService: number;
+  isLoading: boolean;
+  participants: ElasticUser[];
 }
 
 export const initialData: CaseUserActionsState = {
+  caseServices: {},
   caseUserActions: [],
-  firstIndexPushToService: -1,
-  lastIndexPushToService: -1,
   hasDataToPush: false,
-  isLoading: true,
   isError: false,
+  isLoading: true,
   participants: [],
 };
 
 export interface UseGetCaseUserActions extends CaseUserActionsState {
   fetchCaseUserActions: (caseId: string) => void;
 }
 
+const getExternalService = (value: string): CaseFullExternalService | null =>
+  parseString(`${value}`);
+
 const getPushedInfo = (
-  caseUserActions: CaseUserActions[]
-): { firstIndexPushToService: number; lastIndexPushToService: number; hasDataToPush: boolean } => {
-  const firstIndexPushToService = caseUserActions.findIndex(
-    cua => cua.action === 'push-to-service'
-  );
-  const lastIndexPushToService = caseUserActions
-    .map(cua => cua.action)
-    .lastIndexOf('push-to-service');
+  caseUserActions: CaseUserActions[],
+  caseConnectorId: string
+): {
+  caseServices: CaseServices;
+  hasDataToPush: boolean;
+} => {
```

Comment:
this function needs an extra close eye and possibly a refactor. `reduce` then `reduce` containing `filter` then` filter` and containing `reduce`....

Reply:
what do you think about that 
```
const getExternalService = (value: string): CaseExternalService | null =>
  convertToCamelCase<CaseFullExternalService, CaseExternalService>(parseString(`${value}`));

export const getPushedInfo = (
  caseUserActions: CaseUserActions[],
  caseConnectorId: string
): {
  caseServices: CaseServices;
  hasDataToPush: boolean;
} => {
  const hasDataToPushForConnector = (connectorId: string) => {
    const userActionsForPushLessServiceUpdates = caseUserActions.filter(
      mua =>
        (mua.action !== 'push-to-service' &&
          !(mua.action === 'update' && mua.actionField[0] === 'connector_id')) ||
        (mua.action === 'push-to-service' &&
          connectorId === getExternalService(`${mua.newValue}`)?.connectorId)
    );
    return (
      userActionsForPushLessServiceUpdates[userActionsForPushLessServiceUpdates.length - 1]
        .action !== 'push-to-service'
    );
  };

  const caseServices = caseUserActions.reduce<CaseServices>((acc, cua, i) => {
    if (cua.action !== 'push-to-service') {
      return acc;
    }
    const externalService = getExternalService(`${cua.newValue}`);
    if (externalService === null) {
      return acc;
    }

    return {
      ...acc,
      ...(acc[externalService.connectorId] != null
        ? {
            [externalService.connectorId]: {
              ...acc[externalService.connectorId],
              ...externalService,
              lastPushIndex: i,
            },
          }
        : {
            [externalService.connectorId]: {
              ...externalService,
              firstPushIndex: i,
              lastPushIndex: i,
              hasDataToPush: hasDataToPushForConnector(externalService.connectorId),
            },
          }),
    };
  }, {} as CaseServices);

  const hasDataToPush =
    caseServices[caseConnectorId] != null ? caseServices[caseConnectorId].hasDataToPush : true;
  return {
    hasDataToPush,
    caseServices,
  };
};
```

Reply:
yes yes 100 times yes